plugins/kubernetes/app/models/kubernetes/deploy_yaml.rb

```diff
@@ -47,6 +50,31 @@ def template
       end
     end
 
+    # process the annotations and do some string substition
+    # on the keys we find in there.  this allows a single template
+    # to be used for deployments to multiple ENVs and deploy groups
+    # also ask the secret service if the key is really there, if it's not
+    # then inform the user and bail out
+    def set_annotations
+      deploy_group = DeployGroup.find(@release_doc_metadata[:deploy_group_id])
+      environment_permalink = deploy_group.environment.permalink
+      deploy_group_permalink = deploy_group.permalink
+      annotations = @template.spec.template.metadata.annotations
+      annotations.to_h.map do |annotation|
+        if annotation.first =~ /secret\//
+          annotation.last.gsub!(/\$ENV/, environment_permalink) if annotation.last =~ /\$ENV/
+          annotation.last.gsub!(/\$DEPLOY_GROUP/, deploy_group_permalink) if annotation.last =~ /\$DEPLOY_GROUP/
+        end
+        begin
+          SecretStorage.read(annotation.last)
+        rescue
```

Comment:
also prefer not using exceptions to control flow, it is slow / noisy

...

ideally resolve all secrets and then do a read_multi to get all secrets in 1 go AND tell the user that `these 5 secrets are missing` instead of bam this 1 is missing ... bam this other one is missing ... bam ... etc